### ISSUE TITLE
Cannot use larastan and phpstan-strict-rules in combination

### ISSUE THREAD

**Issue by nagmat84** [nunomaduro/larastan#1272]
It is impossible to use Larastan and PhpStan-Strict-Rules in combination. If one does, PhpStan reports a lot of false positives like 

```
Dynamic call to static method error on <eloquent query class here>
```

and similar. The problem has already been reported once in the the old bug https://github.com/nunomaduro/larastan/issues/293. This bug has been closed as "completed", but actually there has never been a proper solution, just a work-around which is to not use PhpStan-Strict-Rules. IMHO this defeats the whole purpose of a static code analyzer to not be strict.

Also note that these false positives disappear, if Larastan is not used, but PhpStan-Strict-Rules is used alone. (Unfortunately, then a lot of other error messages due to Laravels magic pop ups.) It is really the combination of both with breaks things.

Let me illustrate the problem with two examples.

Example 1 (no Laravel magic):

```php
ModelClass::query()->where('column1', '=', 'something')->orderBy('column2')->get();
```

Example 2 (with Laravel magic):

```php
ModelClass::where('column1', '=', 'something')->orderBy('column2')->get();
```

The methods query methods `where`, `orderBy`, `get` and many more are actual instance methods of `Eloquent\Builder` and not defined statically on the model class. Hence, the 1st example is the only one which is correct in a strict sense. The shortcut in the 2nd example works due to Laravel's magic `__callStatic` which forwards non-defined static methods to a freshly created instance of the query builder.

Now, lets discuss what happens for both examples with respect to the different rule sets.

|                                                        | Example 1 (no magic)              | Example 2 (with magic) |
|:---------------------------------------------|:-----------------------------------------:|:------------------------------:|
| PhpStan-Strict-Rules only                |  No error (but see note 1)  | Error: `where` is not a static method of `ModelClass` |
| Larastan only                                  | No error (but see note 2) | No error (but see note 2) |
| Both | False positive: Dynamic call to static method  | False positive: Dynamic call to static method |

Note 1: With PhpStan-Strict-Rules only, one actually gets the best result with respect to the examples. The error for example 2 is completely correct. This would actually tempt me not to use Larastan at all if there weren't a thousand other error messages without it.

Note 2: Even though there are no errors with Larastan only for these two examples, using Larastan only is not a feasible option, because this way a lot of other (and real bugs) are missed due to "sloppiness".

With the combination of both rule sets one gets the aforementioned false positive. If I am allowed to guess, I assume the following happens:

Larastan lies about the methods to PhpStan, re-declares those methods as static class methods and therewith accidentally makes PhpStan forget that these methods are still valid instance methods. Without the strict rules, it is no error to call a static method as a instance method. That might be why we don't see any errors with Larastan only.

To fix the false positive, Larastan should declare the methods as statically accessible _in addition_, but _still keep_ the instance methods.

**Comment by szepeviktor** [comment]
Hello @nagmat84!
Thank you for your highly detailed issue.

We are working hard on Larastan. Larastan does not cover the whole framework (yet).
`__callStatic` and alike are part of Laraval framework to ease the job of developers, `phpstan/phpstan-strict-rules` has another opinion!! It's title says _Extra strict and opinionated rules for PHPStan_
So your problem is Laravel Framework vs. `phpstan/phpstan-strict-rules`

We welcome your contributions!

**Comment by canvural** [comment]
Hi,

This is a known issue. There is already an issue to track the progress.Here: https://github.com/phpstan/phpstan-strict-rules/issues/140

So please follow that issue for the progress.

**Comment by szepeviktor** [comment]
> known issue.

Opened by me!

# 🤯

**Comment by nagmat84** [comment]
While the issue which has been reported at https://github.com/phpstan/phpstan-strict-rules/issues/140 might be somehow related, I am not sure if they share the same cause.

In my specific case I have a proper `Eloquent\Builder` object and calling the method `orderBy` on the object reports a false positive. This method has not been added as a "macro" to the class. it is properly defined inside the class itself. Moreover, it would even be completely wrong to call this method statically, because this method uses`$this` internally:

In other words

```php
$query = MyModel::query();
$query->orderBy('something'); // raises false positive
```

raises a false positive, but the alternative


```php
$query = MyModel::query();
$query::orderBy('something'); // crashes
```

would simply crash. Hence, `orderBy` does not fall into this category of methods which are available statically and dynamically.

**Comment by canvural** [comment]
@nagmat84 I explained in my [comment](https://github.com/phpstan/phpstan-strict-rules/issues/140#issuecomment-959827777) why we are doing that. If you can find a better way, please feel free to send a PR!

**Comment by philbates35** [comment]
If it helps, its possible to still use `phpstan-strict-rules` with larastan, we just need to disable the `strictCalls` rule:

```
parameters:
    paths:
        - app
        - bootstrap
        - config
        - database
        - routes
        - tests
    level: 9
    excludePaths:
        - bootstrap/cache/
    strictRules:
        strictCalls: false
```

**Comment by etrinh** [comment]
Or instead disabling strictCalls (which may avoid reporting other useful warnings), just a regex to this specific one:
```
parameters:
    ...
    ignoreErrors:
        - '#Dynamic call to static method Illuminate\\.*#'
```
